### PR TITLE
Revisit #4976: Harden `quake.phase` lowering and canonicalization

### PR DESCRIPTION
**Pull request by cabreraam**
## Summary

This PR updates `quake.phase` representation, canonicalization, basis conversion, and lowering safety.

The main functional change is the alias-safety check in the anchored `LowerPhase` fallback. The rest is cleanup, clearer canonicalization structure, and documentation of the intended phase lifecycle.

## Review follow-ups
- Added the `Phase(φ) = exp(iφ) I` matrix representation and a circuit symbol in `QuakeOps.td` / `QuakeOps.cpp`

- Updated the `LowerPhase` pass description in `Passes.td` to clarify behavior
  
- Updated the phase canonicalization logic in `CanonicalPatterns.inc` to normalize control polarities and compare controls/polarities with `llvm::zip`

- Reworked `isDirectlyThreadedPhase` in `CanonicalPatterns.inc` to compare relevant threaded wire flow directly

- construct π and 2π with `APFloat` arithmetic rather than using a raw decimal literal

- Added `AdjustAdjointPhasePattern` and registered it from `PhaseOp::getCanonicalizationPatterns` in `QuakeOps.cpp`. This breaks up adjoint handling from the `MergeAdjacentPhasesPattern`

- Removed the unused generic rotation-merge template after `MergeAdjacentPhasePattern` in `CanonicalPatterns.inc`. 

- Added `PhaseOp` to the `QUANTUM_OPS` macro in `QuakeOps.cpp` and removed its separate effects callback. 

- Clarified the `BasisTarget` handling in `DecompositionPatternSelection.cpp`. `quake.phase` remains legal through basis conversion because it is exact-phase bookkeeping, not a physical basis gate to decompose at that point.

- Updated `getWireResultTypes` in `QuakeOps.h` / `QuakeOps.cpp`. The helper no longer receives an `OpBuilder` or constructs a new `!quake.wire` type. It builds the result-type list required by `Op::create` by copying the types of the wire operands themselves, in controls-then-targets order.

- **Alias checks: cascades, block arguments, and `subveq`**
  - `LowerPhase.cpp` now uses `mayAliasForPhaseAnchorFallback` instead of the earlier shallow `mayPhaseAnchorAliasControl` check when it needs the anchored `r1/rz` fallback.
  - The helper in `PhaseUtilities.h` follows identity-preserving def-use chains: `unwrap`, `wrap_new`, control conversions, `extract_ref`, `relax_size`, `subveq`, aggregate operations, and known threaded-wire flow. A non-self `quake.wrap` is also treated as a possible reference rebinding.
  - `subveq` is followed back to its input vector. The analysis intentionally does not prove slices disjoint from their bounds, so slices of the same source remain conservatively treated as potentially overlapping.
  - The analysis stops at block arguments instead of tracing arbitrary CFG predecessors. Function inputs are treated as caller-provided, while non-entry block arguments and opaque values are treated as unknown. If safety cannot be proven, the fallback diagnoses rather than emitting an `r1/rz` sequence with a possible control/target overlap.
  - Added coverage in `lower_phase-errors.qke` for `subveq`, threaded wires, `wrap`, `wrap_new`, function arguments, and CFG block arguments. `lower_phase.qke` covers proven-fresh anchors, including a threaded fresh wire and a self-wrap.

### CODE REVIEW

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h [hunk 1]
```diff
@@ -139,6 +150,179 @@ mayPhaseAnchorAliasControl(const cudaq::quake::StaticQubitTarget &anchor,
...
+  collectPhaseAnchorFallbackRoots(anchor, anchorRoots, at);
+  collectPhaseAnchorFallbackRoots(control, controlRoots, at);
+
+  for (const PhaseAnchorFallbackRoot &anchorRoot : anchorRoots)
+    for (const PhaseAnchorFallbackRoot &controlRoot : controlRoots) {
+      if (anchorRoot.value == controlRoot.value ||
+          anchorRoot.kind == PhaseAnchorFallbackRootKind::Unknown ||
+          controlRoot.kind == PhaseAnchorFallbackRootKind::Unknown ||
+          (anchorRoot.kind == PhaseAnchorFallbackRootKind::FunctionInput &&
+           controlRoot.kind == PhaseAnchorFallbackRootKind::FunctionInput))
```
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: A kernel takes a qubit as an argument, and that kernel is controlled on a qvector whose size is not known at compile time. That produces a `quake.phase` whose anchor and control vector are both function arguments. The anchored `R1/Rz` fallback is the only way to lower it, and the new check refuses.

Tried this repro.qke file.

```
func.func @arg_anchor_with_arg_vector(%theta: f64,
                                      %controls: !quake.veq<?>,
                                      %anchor: !quake.ref) {
  quake.phase (%theta) [%controls] %anchor
      : (f64, !quake.veq<?>, !quake.ref) -> ()
  return
}
```

Output is
```
repro.qke:4:3: error: 'quake.phase' op cannot lower with an anchor that aliases a control operand
  quake.phase (%theta) [%controls] %anchor
  ^
repro.qke:4:3: note: see current operation: "quake.phase"(%arg0, %arg1, %arg2) <{operandSegmentSizes = array<i32: 1, 1, 1>}> : (f64, !quake.veq<?>, !quake.ref) -> ()`
```

Tried exp_pauli below as well with the same error.

```
func.func @dynamic_vector_control_with_separate_anchor(%theta: f64,
                                                       %controls: !quake.veq<?>,
                                                       %anchor: !quake.ref) {
  quake.exp_pauli<adj> (%theta) [%controls neg [true]] %anchor to "I"
      : (f64, !quake.veq<?>, !quake.ref) -> ()
  return
}

```

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h [hunk 2]
```diff
@@ -1,3 +1,11 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
 /*******************************************************************************
```
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: Duplicate copy right header.

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc [hunk 3]
```diff
@@ -789,20 +789,23 @@ static bool haveMatchingPhasePolarities(cudaq::quake::PhaseOp first,
...
+  auto firstPolarities = cudaq::quake::getControlPolarities(
+      firstControls, first.getNegatedQubitControls());
+  auto secondPolarities = cudaq::quake::getControlPolarities(
+      secondControls, second.getNegatedQubitControls());
+  for (auto [firstPolarity, secondPolarity] :
+       llvm::zip(firstPolarities, secondPolarities))
+    if (firstPolarity != secondPolarity)
       return false;
-  }
   return true;
```
Comment (sacpis) [inline cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc]: nit: can be written as `return llvm::equal(firstPolarities, secondPolarities);`.

File: cudaq/test/Transforms/phase_canonicalize.qke [hunk 4]
```diff
@@ -122,10 +122,23 @@ func.func @merge_reference(%a: f64, %b: f64, %control: !quake.ref,
...
+// An adjoint phase canonicalizes to a non-adjoint phase with a negated angle.
+// CHECK-LABEL: func.func @normalize_adjoint_reference(
+// CHECK-SAME: %[[A:[a-zA-Z0-9_.$-]+]]: f64, %[[CONTROL:[a-zA-Z0-9_.$-]+]]: !quake.ref, %[[ANCHOR:[a-zA-Z0-9_.$-]+]]: !quake.ref
+// CHECK: %[[NEG:.*]] = arith.negf %[[A]] : f64
+// CHECK: quake.phase (%[[NEG]]) {{\[}}%[[CONTROL]]{{\]}} %[[ANCHOR]]
+// CHECK-NOT: quake.phase<adj>
+func.func @normalize_adjoint_reference(%a: f64, %control: !quake.ref,
+                                       %anchor: !quake.ref) {
+  quake.phase<adj> (%a) [%control] %anchor
+      : (f64, !quake.ref, !quake.ref) -> ()
```
Comment (sacpis) [inline cudaq/test/Transforms/phase_canonicalize.qke:134]: Should we add a test case for an adjoint phase with negated controls?
```
quake.phase<adj> (%a) [%c neg [true]] %anchor
```
and CHECK that `neg [true]` survives?

**Review by sacpis — COMMENTED** [review]
Overall LGTM. Thanks @cabreraam. Left few comments.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 2 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: Done!

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc (hunk 3 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc]: done!

File: cudaq/test/Transforms/phase_canonicalize.qke (hunk 4 above)
Comment (cabreraam) [reply cudaq/test/Transforms/phase_canonicalize.qke:134]: [added!](https://github.com/NVIDIA/cuda-quantum/pull/5322/changes/36f9da0f86a7fcc44575c6b2fd09df588fa59998)

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 1 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: Thanks for this! I relaxed the check to allow for this case of input arguments of dynamically-sized control `veq` and `!quake.ref` anchor to be lowerable. 

I also modified the test in `ExpPauli-boundaries.qke` to test the same case but for `exp_pauli`

[Relevant commit](https://github.com/NVIDIA/cuda-quantum/pull/5322/changes/283c4f78ce3fd520458bd3ad56b0d6f647833204)

File: cudaq/test/Transforms/DecompositionPatterns/exp_pauli_boundaries.qke [hunk 5]
```diff
@@ -148,8 +148,9 @@ func.func @dynamic_vector_control_with_separate_anchor(%theta: f64,
 }
```
Comment (schweitzpgi) [inline cudaq/test/Transforms/DecompositionPatterns/exp_pauli_boundaries.qke:148]: Let's change the name of this test (and any others) to use the conventions at hand. (I've noticed the agents *love* to do their own thing.)

The file naming convention is snake_case followed optionally by `-` integer or `-error` throughout this directory.

Side note: we also discourage creating deep directory trees so that every test can be called `test.qke` but is buried 17 levels deep in the tree. heh

File: cudaq/include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h [hunk 6]
```diff
@@ -273,10 +273,9 @@ expandKnownSizedControlVeqs(mlir::OpBuilder &builder, mlir::Location location,
                             mlir::ValueRange controls,
                             llvm::ArrayRef<bool> polarities);
 
-/// Return the wire result types for a Quake operator with the given controls
-/// and targets. Quake orders wire results by controls first, then targets.
-llvm::SmallVector<mlir::Type> getWireResultTypes(mlir::OpBuilder &builder,
-                                                 mlir::ValueRange controls,
+/// Return the types of a Quake operator's wire operands in wire-result order:
+/// controls first, then targets.
+llvm::SmallVector<mlir::Type> getWireResultTypes(mlir::ValueRange controls,
```
Comment (schweitzpgi) [inline cudaq/include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h]: I think the convention is to use the re-exports from the `mlir` namespace. So `mlir::SmallVector` and `mlir::ArrayRef` are more the pattern.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h [hunk 7]
```diff
@@ -139,6 +142,178 @@ mayPhaseAnchorAliasControl(const cudaq::quake::StaticQubitTarget &anchor,
...
+  if (auto concat = value.getDefiningOp<cudaq::quake::ConcatOp>()) {
+    for (mlir::Value member : concat.getTargets())
+      collectPhaseAnchorFallbackRoots(member, roots, at);
+    return;
+  }
+  if (auto struq = value.getDefiningOp<cudaq::quake::MakeStruqOp>()) {
+    for (mlir::Value member : struq.getVeqs())
+      collectPhaseAnchorFallbackRoots(member, roots, at);
+    return;
+  }
```
Comment (schweitzpgi) [inline cudaq/lib/Optimizer/Transforms/PhaseUtilities.h:259]: Looks like something that ought to be generated rather than a case tree maintained by hand. (But I didn't look super closely...)

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 1 above)
Comment (schweitzpgi) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: I don't see anything the compiler can legitimately assert on as "aliasing" in Sachin's example. The semantics of the IR are: veq arguments and ref arguments do not alias when used as arguments to the same quantum op  _unless_ one can statically prove they do alias. That's true whether the number of refs in the veq is statically known or not.

The converse cannot be true otherwise it would invalidate all manner of legitimate programs that CUDA-Q supports. That is, we must be able to thread veq arguments and ref arguments and assume that the programmer knows what they are doing to support legal programs. Yes, this opens the door that the programmer does _not_ know what they are doing and shoots themselves in the foot. That's on them.

Comment (schweitzpgi) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: This is a rather vivid illustration actually of why references and values are distinct. Full stop.

In value semantics, we want to have _wire identities_ threading into Ops. As identities, we can ascertain immediately if the same qubit arrives in multiple input positions. Better yet, they are linear types that obey SSI and we cannot even construct that IR. That is by design.

With references, you enter the roller coaster with no safety bar. References mean aliasing is possible. But that shouldn't be construed to mean aliasing is ubiquitous everywhere. So while a reference can refer to any logical qubit by design, we do have to have some tolerance and expect the user to write quantum circuits where not every ref aliases with every other ref.

So that means,
```c++
  cudaq::qvector v(4);
  x<cudaq::ctrl>(v, v[0]);
```
is valid input, but it contains a logical error. Depending on how this gets lowered, this logical error will become manifest when the circuit is splayed out, or at code gen time, or possibly by the simulator target which might catch the error at its own higher-level representation.

File: cudaq/test/Transforms/DecompositionPatterns/exp_pauli_boundaries.qke (hunk 5 above)
Comment (cabreraam) [reply cudaq/test/Transforms/DecompositionPatterns/exp_pauli_boundaries.qke:148]: Done!

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 7 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h:259]: I will address this in a future PR. Thanks for flagging!

File: cudaq/include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h (hunk 6 above)
Comment (cabreraam) [reply cudaq/include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h]: Addressed in https://github.com/NVIDIA/cuda-quantum/pull/5322/commits/54d062702dd0df44859998cfccce8d2209a24d94

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc [hunk 8]
```diff
@@ -1191,7 +1211,15 @@ struct MergeAdjacentPhasePattern
 
   LogicalResult matchAndRewrite(cudaq::quake::PhaseOp phase,
                                 PatternRewriter &rewriter) const override {
-    auto *previous = phase->getPrevNode();
+    Operation *previous = phase->getPrevNode();
```
Comment (schweitzpgi) [inline cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc:1214]: Isn't this looking at the previous op in the intrusive linked list structure? How do you know that's even used by `phase`?

Comment (schweitzpgi) [reply cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc:1214]: Why wouldn't we look at `getParameter()`'s `Value` and use `getDefiningOp<arith::NegFOp>()` here? It avoids trying to make a lucky guess.

I also wonder about pattern matching against `NegFOp`. The compiler can fold that op away rather easily.

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc [hunk 9]
```diff
@@ -1201,26 +1229,16 @@ struct MergeAdjacentPhasePattern
...
-      adjoint = input.getIsAdjAttr();
-      mergedAngle =
-          arith::SubFOp::create(rewriter, loc, firstAngle, secondAngle);
-    }
+    auto mergedAngle =
+        arith::AddFOp::create(rewriter, loc, firstAngle, secondAngle);
 
     rewriter.replaceOpWithNewOp<cudaq::quake::PhaseOp>(
-        phase, phase.getResultTypes(), adjoint, ValueRange{mergedAngle},
+        phase, phase.getResultTypes(), UnitAttr{}, ValueRange{mergedAngle},
```
Comment (schweitzpgi) [inline cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc:1241]: Does PhaseOp have builders so we don't have to supply the default dummy arguments inline? (i.e. `UnitAttr{}`)

File: cudaq/lib/Optimizer/Transforms/LowerPhase.cpp
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/LowerPhase.cpp:186]: Seems like `lower-phase` now rejects a valid kernel when the phase anchor is a block argument? It fails when a phase with an unresolved-size veq control whose anchor wire is carried by a loop block argument. The wire below comes from `quake.null_wire`, so it cannot be in `%controls`, yet the pass fails

Here is the repro.qke
```
func.func @loop_wire_anchor(%theta: f64, %controls: !quake.veq<?>, %n: i64) {
  %w0 = quake.null_wire
  %c0 = arith.constant 0 : i64
  %c1 = arith.constant 1 : i64
  cf.br ^body(%c0, %w0 : i64, !quake.wire)
^body(%i: i64, %w: !quake.wire):
  %r = quake.phase (%theta) [%controls] %w : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
  %inext = arith.addi %i, %c1 : i64
  %done = arith.cmpi slt, %inext, %n : i64
  cf.cond_br %done, ^body(%inext, %r : i64, !quake.wire), ^exit(%r : !quake.wire)
^exit(%wend: !quake.wire):
  quake.sink %wend : !quake.wire
  return
}
```

gives an error as
```
repro.qke:7:8: error: 'quake.phase' op cannot lower with an anchor that aliases a control operand
  %r = quake.phase (%theta) [%controls] %w : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
       ^
repro.qke:7:8: note: see current operation: %5 = "quake.phase"(%arg0, %arg1, %4) <{operandSegmentSizes = array<i32: 1, 1, 1>}> : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
```

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc (hunk 8 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc:1214]: Okay, I took your suggestion to rework the logic to use the `getDefiningOp` on the phase angle argument instead of guessing. Thanks! 

I looked into `NegFOp` folding as well and I think pattern matching against it should be fine. The point of the checking for a preceding negated angle to a phase op is just to ignore it. So something like a double negation would just be erased, so folding would just ignore that bit of safety logic.

I think it does raise an interesting point about other arithmetic operations that might find their way in between phases, but this is accounted for in the `NormalizePhasePlacement` pass, rather than this canonicalization peephole.

File: cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc (hunk 9 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Dialect/Quake/CanonicalPatterns.inc]: Good point. I _am_ keeping the supplying of the default dummy argument for this case (`is_adj=false`), but I am using the builder that takes a boolean for this parameter and explicitly naming it in the function call. I've seen this convention in other parts of the code so I'm going to use it here.

File: cudaq/lib/Optimizer/Transforms/LowerPhase.cpp
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/LowerPhase.cpp:187]: Okay, I have a fix for this now. Thanks for pointing this out.

What I do now is just add an [additional check for block arguments](https://github.com/NVIDIA/cuda-quantum/pull/5322/changes/1a70f61050b6ec5de3ec4a334e3ef79a691dafb1#diff-bca59250ac8deee0a138f3f02993c22b610c936b1b5af8c4eb2612fee6b89449R291-R339) when trying to find the roots of `phase` controls and anchors. In that code, here are the checks that occur
- if the terminator of the block containing the block argument is not a branch op, we conservatively say that the root type of the block argument is unknown
- if it is a branch op, we trace all of the branch operands the same recursive way by calling `collectPhaseFallbackRootsImpl`
- if there is a loop backedge from one of the branch operands, we will have added it to a set. In a future iteration, if we've seen it before, we can safely ignore it because we'll have accounted for it once already

Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/LowerPhase.cpp:187]: Instead of too conservatively blocking phase lowering, your repro now emits
```mlir
  func.func @loop_wire_anchor(%arg0: f64, %arg1: !quake.veq<?>, %arg2: i64) {
    %0 = quake.null_wire
    %c0_i64 = arith.constant 0 : i64
    %c1_i64 = arith.constant 1 : i64
    cf.br ^bb1(%c0_i64, %0 : i64, !quake.wire)
  ^bb1(%1: i64, %2: !quake.wire):  // 2 preds: ^bb0, ^bb1
    %3 = arith.negf %arg0 : f64
    %4 = quake.r1 (%arg0) [%arg1] %2 : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
    %5 = quake.r1 (%arg0) [%arg1] %4 : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
    %6 = quake.rz (%3) [%arg1] %5 : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
    %7 = quake.rz (%3) [%arg1] %6 : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
    %8 = arith.addi %1, %c1_i64 : i64
    %9 = arith.cmpi slt, %8, %arg2 : i64
    cf.cond_br %9, ^bb1(%8, %7 : i64, !quake.wire), ^bb2(%7 : !quake.wire)
  ^bb2(%10: !quake.wire):  // pred: ^bb1
    quake.sink %10 : !quake.wire
    return
  }
```
where we have fallen into the fallback case where no scalar anchor from the controls of the original `phase` op can be identifies. So in this case, we can now determine that the phase op target and controls _don't_ alias now that the logic to track down block argument provenance has been added. In this case, %w is shown to be freshly created in the block preceding `%^bb1` so the phase op _is_ lowered.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h [hunk 10]
```diff
@@ -139,6 +144,249 @@ mayPhaseAnchorAliasControl(const cudaq::quake::StaticQubitTarget &anchor,
...
+    return true;
+
+  bool mayBeRebound = false;
+  function.walk([&](cudaq::quake::WrapOp wrap) {
+    if (!phaseValuesMayShareRoot(root, wrap.getRefValue()) ||
+        isProvenSelfPhaseWrap(wrap))
+      return;
+    if (wrap->getBlock() != at->getBlock() || wrap->isBeforeInBlock(at))
+      mayBeRebound = true;
+  });
```
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: nit: Seems like the alias check rescans the whole function for every question? Should we collect the non-self quake.wrap ops once per function into a small map keyed by root, and reuse it. And then add WalkResult::interrupt() on the first hit. That way we can exit early.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h [hunk 11]
```diff
@@ -139,6 +144,249 @@ mayPhaseAnchorAliasControl(const cudaq::quake::StaticQubitTarget &anchor,
...
+        if (value == result)
+          return collectPhaseAnchorFallbackRootsImpl(flow->inputs[index], roots,
+                                                     def, visitedWireArguments);
+
+    if (mlir::isa<cudaq::quake::AllocaOp>(def)) {
+      roots.push_back({value, mayHaveReboundPhaseRoot(value, at)
+                                  ? PhaseAnchorFallbackRootKind::Unknown
+                                  : PhaseAnchorFallbackRootKind::FreshLocal});
+      return;
+    }
```
Comment (sacpis) [inline cudaq/lib/Optimizer/Transforms/PhaseUtilities.h:304]: Seems like a phase inside the loop does not compile.

```
func.func @k(%theta: f64, %size: i32) {
  %q = quake.alloca !quake.ref
  %ctrl = quake.alloca !quake.veq<?>[%size : i32]
  %c0 = arith.constant 0 : i64
  %c4 = arith.constant 4 : i64
  %c1 = arith.constant 1 : i64
  cc.loop while ((%i = %c0) -> (i64)) {
    %cond = arith.cmpi slt, %i, %c4 : i64
    cc.condition %cond (%i : i64)
  } do {
   ^bb0(%i: i64):
    quake.phase (%theta) [%ctrl] %q : (f64, !quake.veq<?>, !quake.ref) -> ()
    cc.continue %i : i64
  } step {
   ^bb0(%i: i64):
    %n = arith.addi %i, %c1 : i64
    cc.continue %n : i64
  }
  return
}
```

with command as 
```
./build/bin/cudaq-opt --memtoreg --lower-phase repro.qke
```

gives output as
```
repro.qke:12:5: error: 'quake.phase' op cannot lower with an anchor that aliases a control operand
    quake.phase (%theta) [%ctrl] %q : (f64, !quake.veq<?>, !quake.ref) -> ()
    ^
repro.qke:12:5: note: see current operation: %7 = "quake.phase"(%arg0, %1, %arg5) <{operandSegmentSizes = array<i32: 1, 1, 1>}> : (f64, !quake.veq<?>, !quake.wire) -> !quake.wire
```

**Review by sacpis — APPROVED** [review]
Overall LGTM. Thanks @cabreraam. Left few comments.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 10 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: Thanks for pointing this out!

I now collect the non-self quake.wrap operations once per function right at the beginning of the LowerPhase pass into a map keyed by root and reuse it for each alias check. When we enter this function, we'll lookup the root to see if there are any associated wrap ops. If there aren't, we know that either there is no self-wrap or there isn't a wrap associated with that !quake.ref, so we can conclude that the ref was not re-bound. (This is basically the WalkResult::interrupt() functionality you were suggesting.) When there is a potential for a rebind, then we fall into the original logic. Addressed in 127e8f182a

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 11 above)
Comment (cabreraam) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h:304]: Thanks for this. 

`memtoreg` threads the anchor wire through `cc.loop` region arguments, but the provenance logic previously handled only ordinary CFG branches. It now follows structured region operands through `RegionBranchOpInterface`, while retaining the existing cycle and unknown-path safeguards. I added both the repro and negative alias coverage. Addressed in da0bf4f483

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 10 above)
Comment (sacpis) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h]: Thanks @cabreraam.

File: cudaq/lib/Optimizer/Transforms/PhaseUtilities.h (hunk 11 above)
Comment (sacpis) [reply cudaq/lib/Optimizer/Transforms/PhaseUtilities.h:304]: Thanks @cabreraam.

File: cudaq/lib/Optimizer/Transforms/LowerPhase.cpp
Comment (sacpis) [reply cudaq/lib/Optimizer/Transforms/LowerPhase.cpp:187]: Thanks @cabreraam.
